chain-signatures/node/src/indexer_eth/mod.rs

```diff
@@ -0,0 +1,1449 @@
+    ) -> anyhow::Result<Option<alloy::rpc::types::Transaction>> {
+        match self {
+            EthereumClient::Helios(client) => client.get_transaction_by_hash(tx_hash).await,
+            EthereumClient::DirectRpc(client) => client.get_transaction_by_hash(tx_hash).await,
+        }
+    }
+
+    pub async fn call(
+        &self,
+        from: Address,
+        to: Address,
+        data: Bytes,
+        block_number: u64,
+    ) -> anyhow::Result<Bytes> {
+        match self {
+            EthereumClient::Helios(client) => client.call(from, to, data, block_number).await,
+            EthereumClient::DirectRpc(client) => client.call(from, to, data, block_number).await,
+        }
+    }
+
+    async fn get_latest_block_number(&self) -> anyhow::Result<u64> {
+        match self {
+            EthereumClient::Helios(client) => client.get_latest_block_number().await,
+            EthereumClient::DirectRpc(client) => client.get_latest_block_number().await,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct EthereumIndexer {
+    eth: EthConfig,
+    sign_tx: mpsc::Sender<Sign>,
+    app_data_storage: AppDataStorage,
+    node_near_account_id: AccountId,
+    backlog: Backlog,
+    contract_watcher: ContractStateWatcher,
+    mesh_state: watch::Receiver<MeshState>,
+    node_client: NodeClient,
+    client: EthereumClient,
+}
+
+impl EthereumIndexer {
+    #[allow(clippy::too_many_arguments)]
+    pub async fn new(
+        eth: Option<EthConfig>,
+        sign_tx: mpsc::Sender<Sign>,
+        app_data_storage: AppDataStorage,
+        node_near_account_id: AccountId,
+        backlog: Backlog,
+        contract_watcher: ContractStateWatcher,
+        mesh_state: watch::Receiver<MeshState>,
+        node_client: NodeClient,
+    ) -> anyhow::Result<Self> {
+        let Some(eth) = eth else {
+            tracing::warn!("ethereum indexer is disabled");
+            return Err(anyhow::anyhow!("ethereum indexer is disabled"));
+        };
+
+        let client = EthereumClient::new(eth.clone()).await?;
+
+        Ok(Self {
+            eth,
+            sign_tx,
+            app_data_storage,
+            node_near_account_id,
+            backlog,
+            contract_watcher,
+            mesh_state,
+            node_client,
+            client,
+        })
+    }
+
+    pub async fn run(self) {
+        let backlog = self.backlog;
+        let contract_watcher = self.contract_watcher;
+        let mesh_state = self.mesh_state;
+        let node_client = self.node_client;
+        let app_data_storage = self.app_data_storage;
+        let client = self.client;
+        let eth = self.eth;
+        let sign_tx = self.sign_tx;
+        let node_near_account_id = self.node_near_account_id;
+
+        Self::recover_backlog(&backlog, contract_watcher, mesh_state, node_client).await;
+
+        let last_processed_block = Self::get_last_processed_block(&app_data_storage).await;
+
+        let client = Arc::new(client);
```

Comment:
this shouldn't be an Arc. All the internal clients are cloneable. The reqwest client is internally reference counted, and you have an Arc over helios client anyways. So this Arc is useless